src/main/java/com/datadoghq/sketch/ddsketch/store/rescaling/RescalingStore.java

```diff
@@ -0,0 +1,125 @@
+package com.datadoghq.sketch.ddsketch.store.rescaling;
+
+import com.datadoghq.sketch.ddsketch.store.Bin;
+import com.datadoghq.sketch.ddsketch.store.Store;
+
+import java.util.Iterator;
+import java.util.function.Supplier;
+
+public abstract class RescalingStore implements Store {
+
+    protected final Supplier<? extends Store> storeSupplier;
+
+    protected Store delegate;
+    protected int scalingFactor = 1;
+
+    public RescalingStore(Supplier<? extends Store> storeSupplier) {
+        this.storeSupplier = storeSupplier;
+        this.delegate = storeSupplier.get();
+    }
+
+    protected RescalingStore(Supplier<? extends Store> storeSupplier, Store delegate, int scalingFactor) {
+        this.storeSupplier = storeSupplier;
+        this.delegate = delegate;
+        this.scalingFactor = scalingFactor;
+    }
+
+    /**
+     * @return the rescaling factor > 1 if it should rescale
+     */
+    protected abstract int shouldRescale();
+
+    /**
+     * Merge contiguous buckets.
+     *
+     * @param rescalingFactor the factor to be compounded with the current one
+     */
+    protected final void rescale(int rescalingFactor) {
+        final Store newDelegate = storeSupplier.get();
+        final int newScalingFactor = scalingFactor * rescalingFactor;
+        delegate.getStream()
+            .map(bin -> new Bin(Math.floorDiv(bin.getIndex(), rescalingFactor), bin.getCount()))
+            .forEach(newDelegate::add);
+        delegate = newDelegate;
+        scalingFactor = newScalingFactor;
+    }
+
+    @Override
+    public void add(int index, double count) {
+        delegate.add(Math.floorDiv(index, scalingFactor), count);
+        final int rescalingFactor = shouldRescale();
+        if (rescalingFactor > 1) {
+            rescale(rescalingFactor);
+        }
+    }
+
+    @Override
+    public void clear() {
+        delegate.clear();
+    }
+
+    @Override
+    public void mergeWith(Store store) {
+        if (!(store instanceof RescalingStore) || delegate.maxShift() > 0) {
+            // FIXME: implement
+            throw new UnsupportedOperationException();
+        }
+        final RescalingStore that = (RescalingStore) store;
+        final int newScalingFactor = lcm(this.scalingFactor, that.scalingFactor);
+        if (newScalingFactor > scalingFactor) {
+            rescale(newScalingFactor / scalingFactor);
+        }
```

Comment:
These values can only be powers of 2 (unless I'm confused) but I doubt the JIT compiler will figure it out and allow these integer divisions to be strength reduced. The rescaling could be rewritten to use shifts instead.

Reply:
Good point, although as of now, `RescalingStore` does not generally impose it to be a power of 2, only `BoundedSizeRescalingStore` does. We can tweak things to make sure we can have that optimization though.